Callers of ContainerNode::removeChild need to keep a strong reference to the child

achristensen07 opened the pull request:
#### 96d8c10caa724a3e35011a782d94e4bb882692d6
<pre>
Callers of ContainerNode::removeChild need to keep a strong reference to the child
<a href="https://bugs.webkit.org/show_bug.cgi?id=252769">https://bugs.webkit.org/show_bug.cgi?id=252769</a>
rdar://105784240

Reviewed by Geoff Garen and Chris Dumez.

260535@main introduced some operations that use the Node after calling
removeNodeWithScriptAssertion which may destroy it.

* Source/WebCore/dom/ContainerNode.cpp:
(WebCore::ContainerNode::removeChild):
* Source/WebCore/dom/ContainerNode.h:

Canonical link: <a href="https://commits.webkit.org/260708@main">https://commits.webkit.org/260708@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/7b2b7a19b30981a04d2764e6b3f0e70347b19f60

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/109143 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/18225 "Built successfully") | [  ~~🛠 mac~~](https://ews-build.webkit.org/#/builders/43/builds/41962 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/87/builds/675 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/118390 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [  ~~🧪 bindings~~](https://ews-build.webkit.org/#/builders/11/builds/113026 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 ios-sim~~](https://ews-build.webkit.org/#/builders/76/builds/19684 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 mac-AS-debug~~](https://ews-build.webkit.org/#/builders/85/builds/9527 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/36/builds/101384 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [  ~~🧪 webkitperl~~](https://ews-build.webkit.org/#/builders/19/builds/114901 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/76/builds/19684 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/43/builds/41962 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/101384 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/76/builds/19684 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/43/builds/41962 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/36/builds/101384 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/81/builds/11005 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/43/builds/41962 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/82/builds/11736 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/85/builds/9527 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/80/builds/17105 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/43/builds/41962 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/79/builds/13352 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/75/builds/4041 "Built successfully and passed tests") | | | | 
<!--EWS-Status-Bubble-End-->

geoffreygaren commented on Source/WebCore/dom/ContainerNode.cpp:
```diff
@@ -659,7 +659,7 @@ void ContainerNode::disconnectDescendantFrames()
 }
 
-ExceptionOr<void> ContainerNode::removeChild(Node& oldChild)
+ExceptionOr<void> ContainerNode::removeChild(const Ref<Node>& oldChild)
```
We want `const Ref<Node>` here. 

`const Ref<Node>&` can be more efficient, but nothing prevents the referenced `Ref<Node>` from changing what it points to in surprising ways. For example, if someone called `removeChild(m_firstChild)`, something bad might happen.

cdumez commented on Source/WebCore/dom/ContainerNode.cpp:
```diff
@@ -659,7 +659,7 @@ void ContainerNode::disconnectDescendantFrames()
 }
 
-ExceptionOr<void> ContainerNode::removeChild(Node& oldChild)
+ExceptionOr<void> ContainerNode::removeChild(const Ref<Node> oldChild)
```
cc @rniwa here but I thought we didn't do this, even with our new safer patterns.

Also, if we decide to do this, we should probably pass a `const Ref<Node>&` instead of a `const Ref<Node>`.

But maybe we should simply have a protector for oldChild inside the implementation?

cdumez replied:
I see that I am contradicting Geoff above. However, passing a `const Ref<Node>&` or a `const Ref<Node>` is just not something we usually do in WebKit as far as I know.

I am curious if this is a new style we want and why we can't use a protector variable like we usually do.

geoffreygaren replied:
@cdumez is right that we don't usually do it this way. Our long term plan for local RefPtr verification would instead verify that the _caller_ held a Ref<Node> or RefPtr<Node> in scope.

Using a protector variable would also work here. An earlier version @achristensen07 showed me did it that way. 

My main feedback is that `const Ref<Node>&` is not completely safe.

If Node& argument and a protector variable are more congruent with other code, I'm cool with that too.

achristensen07 replied:
I'm happy to do any of the 3 things which are mostly equivalent in this case if there's a consensus among reviewers

geoffreygaren replied:
OK, let's do local variable protector, to be similar to other code.

geoffreygaren reviewed (approved):
r=me

cdumez commented on Source/WebCore/dom/ContainerNode.cpp:
```diff
@@ -665,7 +665,8 @@ ExceptionOr<void> ContainerNode::removeChild(Node& oldChild)
 
-    Ref<ContainerNode> protectedThis(*this);
+    Ref protectedThis { *this };
+    Ref protectedOldNode { oldChild };
```
protectedOldChild ?

Commit:
Callers of ContainerNode::removeChild need to keep a strong reference to the child
https://bugs.webkit.org/show_bug.cgi?id=252769
rdar://105784240

Reviewed by Geoff Garen and Chris Dumez.

260535@main introduced some operations that use the Node after calling
removeNodeWithScriptAssertion which may destroy it.

* Source/WebCore/dom/ContainerNode.cpp:
(WebCore::ContainerNode::removeChild):
* Source/WebCore/dom/ContainerNode.h:

Canonical link: https://commits.webkit.org/260708@main
